### PR TITLE
add new FMI config Flags to filter ModelDescription Vars

### PR DESCRIPTION
**Pull request by arun3688**
### Purpose
This provides functionality to filter FMI- ModelDescription vars in modelDescription.xml with options similar to Dymola. The valid options are
1. none
2. internal  // default
3. protected
4. blackBox

### Related issue 
see https://trac.openmodelica.org/OpenModelica/ticket/3105

### changedFiles
Flags.mo -- added new Config flags FMI_FILTER
FlagUtil.mo -- added new Config flags FMI_FILTER
SimCodeUtil.mo  -- Filtered the variables according to different Flags

### usage
```
omc --fmiFilter = blackBox  a.mos 
omc --fmiFilter = protected  a.mos

```

### CODE REVIEW

File: OMCompiler/Compiler/Util/Flags.mo [hunk 1]
```diff
@@ -160,6 +160,13 @@ public constant Integer METAMODELICA = 2;
 public constant Integer PARMODELICA = 3;
 public constant Integer OPTIMICA = 4;
 public constant Integer PDEMODELICA = 5;
+
+// FMI-ModelDescription-ENUM-FLAGS
+public constant Integer FMI_NONE = 6;
```
Comment (lochel) [inline OMCompiler/Compiler/Util/Flags.mo:165]: This should start at 1, even though it actually doesn't matter at all.

File: OMCompiler/Compiler/Util/Flags.mo [hunk 2]
```diff
@@ -1348,6 +1353,10 @@ constant ConfigFlag FMI_VERSION = CONFIG_FLAG(140,
 constant ConfigFlag FLAT_MODELICA = CONFIG_FLAG(141, "flatModelica",
   SOME("f"), EXTERNAL(), BOOL_FLAG(false), NONE(),
   Gettext.gettext("Outputs experimental flat Modelica."));
+constant ConfigFlag FMI_FILTER = CONFIG_FLAG(142, "fmiFilter", NONE(), EXTERNAL(),
+  ENUM_FLAG(FMI_INTERNAL, {("none", FMI_NONE), ("internal", FMI_INTERNAL), ("protected", FMI_PROTECTED), ("blackBox", FMI_BLACKBOX)}),
+  SOME(STRING_OPTION({"none", "internal", "protected", "blackBox"})),
+  Gettext.gettext("Filters the FMI-ModelDescription Vars in the ModelDescription.xml"));
```
Comment (lochel) [inline OMCompiler/Compiler/Util/Flags.mo:1359]: We need to explain this in the user's guide ...

File: testsuite/openmodelica/fmi/ModelExchange/2.0/fmi_attributes_11.mos [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+// "fmi_attributes_11.fmu"
+// ""
+// 0
+// ""
+// 0
+// ""
+// "  <ModelVariables>
+//   <!-- Index of variable = \"1\" -->
+//   <ScalarVariable
+//     name=\"x\"
```
Comment (lochel) [inline testsuite/openmodelica/fmi/ModelExchange/2.0/fmi_attributes_11.mos:49]: All names (except for inputs/outputs) should be concealed.

Comment (kabdelhak) [reply testsuite/openmodelica/fmi/ModelExchange/2.0/fmi_attributes_11.mos:49]: How would you do that? Just create unique cryptic names?

Comment (arun3688) [reply testsuite/openmodelica/fmi/ModelExchange/2.0/fmi_attributes_11.mos:49]: In Dymola they are using some thing like , concealed1, concealed2, just a random unique name

File: OMCompiler/Compiler/Util/Flags.mo (hunk 1 above)
Comment (arun3688) [reply OMCompiler/Compiler/Util/Flags.mo]: updated the values to start from 1

File: testsuite/openmodelica/fmi/ModelExchange/2.0/fmi_attributes_11.mos (hunk 3 above)
Comment (arun3688) [reply testsuite/openmodelica/fmi/ModelExchange/2.0/fmi_attributes_11.mos]: this is implemented now see https://github.com/OpenModelica/OpenModelica/pull/880/commits/e932328bea48c761af84cb37ee2a30bb5afde67f

File: OMCompiler/Compiler/SimCode/SimCodeUtil.mo [hunk 4]
```diff
@@ -2828,9 +2828,9 @@ algorithm
     case(DAE.CREF(cr, ty)::rest) equation
       slst = List.map(dims, intString);
       if FMI.isFMIVersion20() then
-        var = SimCodeVar.SIMVAR(cr, BackendDAE.VARIABLE(), "", "", "", 0, NONE(), NONE(), NONE(), NONE(), false, ty, false, SOME(name), SimCodeVar.NOALIAS(), DAE.emptyElementSource, SOME(SimCodeVar.LOCAL()), NONE(), NONE(), slst, false, true, true, NONE(), NONE(), NONE(), NONE(), true);
+        var = SimCodeVar.SIMVAR(cr, BackendDAE.VARIABLE(), "", "", "", 0, NONE(), NONE(), NONE(), NONE(), false, ty, false, SOME(name), SimCodeVar.NOALIAS(), DAE.emptyElementSource, SOME(SimCodeVar.LOCAL()), NONE(), NONE(), slst, false, true, true, NONE(), NONE(), NONE(), NONE(), NONE());
```
Comment (lochel) [inline OMCompiler/Compiler/SimCode/SimCodeUtil.mo:2831]: Why did you change `true` to `NONE()`?

File: OMCompiler/Compiler/Template/SimCodeTV.mo [hunk 5]
```diff
@@ -270,7 +270,7 @@ package SimCodeVar
       Option<String> matrixName;
       Option<Variability> variability "FMI-2.0 variabilty attribute";
       Option<Initial> initial_ "FMI-2.0 initial attribute";
-      Boolean exportVar "true for internal variables that are introduced in the symbolic transformation process";
+      Option<DAE.ComponentRef> exportVar "variables will only be exported to the modelDescription.xml if this attribute is SOME(cref)";
```
Comment (lochel) [inline OMCompiler/Compiler/Template/SimCodeTV.mo:273]: state also that cref is only used in the modelDEscription.xml

File: OMCompiler/Compiler/SimCode/SimCodeUtil.mo (hunk 4 above)
Comment (arun3688) [reply OMCompiler/Compiler/SimCode/SimCodeUtil.mo]: Changed true to SOME(cref) and false to NONE() see https://github.com/OpenModelica/OpenModelica/pull/880/commits/219bb4351f89a4d5926553cbe251899648011004

File: OMCompiler/Compiler/Template/SimCodeTV.mo (hunk 5 above)
Comment (arun3688) [reply OMCompiler/Compiler/Template/SimCodeTV.mo:273]: added the information

**Comment by lochel** [comment]
There might be some tests that need be updated through.

File: OMCompiler/Compiler/Util/Flags.mo [hunk 6]
```diff
@@ -553,8 +560,6 @@ constant DebugFlag NF_DUMP_FLAT = DEBUG_FLAG(192, "nfDumpFlat", false,
   Gettext.gettext("Dumps the flat model structure before generating the DAE."));
 constant DebugFlag DUMP_FORCE_FMI_ATTRIBUTES = DEBUG_FLAG(193, "force-fmi-attributes", false,
   Gettext.gettext("Force to export all fmi attributes to the modelDescription.xml, including those which have default values"));
-constant DebugFlag DUMP_FORCE_FMI_INTERNAL_VARIABLES = DEBUG_FLAG(194, "force-fmi-internal-variables", false,
-  Gettext.gettext("Force to export all internal variables (eg: $CSE) to the modelDescription.xml"));
```
Comment (lochel) [inline OMCompiler/Compiler/Util/Flags.mo:563]: Didn't we test this flag? I think we need test for `none` and `internal` as well.

Comment (lochel) [reply OMCompiler/Compiler/Util/Flags.mo:563]: @arun3688 Please open a separate PR for this issue. I will merge this now.

**Comment by adeas31** [comment]
We made these options exclusive. In Dymola you can select all of them.

**Comment by lochel** [comment]
This is on purpose. We didn't implement the same options that Dymola supports. What combination of our options would make sense in your opinion?

**Comment by adeas31** [comment]
I saw [ticket#3105#comment:4](https://trac.openmodelica.org/OpenModelica/ticket/3105#comment:4) which makes it clear. Thanks!

File: OMCompiler/Compiler/Util/Flags.mo (hunk 6 above)
Comment (arun3688) [reply OMCompiler/Compiler/Util/Flags.mo:563]: @lochel added test cases for checking fmiFilter= none and internal see https://github.com/OpenModelica/OpenModelica/pull/889
